### ISSUE TITLE
model.py refactoring: create transformer+MLP blocks for the engines

### ISSUE THREAD

**Issue by kctezcan** [ecmwf/WeatherGenerator#130]
### Is your feature request related to a problem? Please describe.

At the current stage the create() function in model.py is creating the embedding network, the local/global engines etc... These are being explicitly created in the code with for loops appending transformer architectures and MLPs. This is reducing the readibility of the code. (see e.g. [here](https://github.com/ecmwf/WeatherGenerator/blob/23435d2b210031cb8fd7a5a06a41c051ab3f7f29/src/weathergen/model/model.py#L261)    )

### Describe the solution you'd like

Instead these explicit for loops of transformers+MLPs should be impemented as classes in attention.py and these should be called in the model.py file

**Comment by kctezcan** [comment]
Hi @clessig,  some points with my suggestions to discuss. Please let me know what you think.

1. sure, we can put the new classes in a new file. However as these blocks can be combinations of MLP, linear, transformer, identity etc layers, I think naming the file transformers.py may be misleading. Instead, we can name the file engines.py or something similar as the file would contain the different local/global assimilation engines. 
2. Now we have multiple files for defining some base model parts, such as
2.1. mlp.py for the MLP (composed of linear layers, layernorm, dropout) (~55 lines of code)
2.2 stream_embed_linear.py for StreamEmbedLinear (composed of a linear layer) (~9 lines of code)
2.3 stream_embed_transformer for StreamEmbedTransformer (composed of MLP, MultiSelfAttentionHead, linear, dropout etc) (~150 lines of code)
2.4 ens_prediction_head.py for EnsPredictionHead (composed of linear layers) (~35 lines of code)
2.5 I think we can put all these into a single file, which we call blocks.py to reduce from 4 files to 1 file. This would contain all the "building blocks"
3. I suggest to make a new file called engines.py and make 4 classes in there to use in the model.py:
3.1 `Class LocalAssimilationEngine()`
3.1.1 For instance here the code below [(line 162-183)](https://github.com/ecmwf/WeatherGenerator/blob/23435d2b210031cb8fd7a5a06a41c051ab3f7f29/src/weathergen/model/model.py#L162) would go into the class:
```python
        # local assimilation engine
        self.ae_local_blocks = torch.nn.ModuleList()
        for _ in range(cf.ae_local_num_blocks):
            self.ae_local_blocks.append(
                MultiSelfAttentionHead_Varlen(
                    cf.ae_local_dim_embed,
                    num_heads=cf.ae_local_num_heads,
                    dropout_rate=cf.ae_local_dropout_rate,
                    with_qk_lnorm=cf.ae_local_with_qk_lnorm,
                    with_flash=cf.with_flash_attention,
                    norm_type=cf.norm_type,
                )
            )
            self.ae_local_blocks.append(
                MLP(
                    cf.ae_local_dim_embed,
                    cf.ae_local_dim_embed,
                    with_residual=True,
                    dropout_rate=cf.ae_local_dropout_rate,
                    norm_type=cf.norm_type,
                )
            )
``` 
3.2 `Class LocalGlobalAdapterEngine()`
3.3 `Class GlobalAssimilationEngine()`
3.4 `Class ForecastingEngine()`
3.5 all these engines only need the config variable `cf` to build the blocks, so their create() function would take only the config as input and return the ModuleList, i.e.
3.5.1 `def create(cf: Config): -> torch.nn.ModuleList()` and `self.ae_local_blocks = LocalAssimilationEngine().create(cf)`
3.6. I would still keep the functions like `assimilate_local()`, `assimilate_global()` etc in the model.py since these require access to the self variables created in the `Model().create()`. We can move these into the respective class in the future as well.
